keras/src/models/model.py

```diff
@@ -556,6 +556,125 @@ def _get_variable_map(self):
         map_saveable_variables(self, store=store, visited_saveables=set())
         return store
 
+    def get_nested_variables(
+        self,
+        trainable_variables=False,
+        non_trainable_variables=False,
+        optimizer_variables=False,
+        metrics_variables=False,
+    ):
+        """Retrieves tree-like structure of specified variables of the model.
+
+        This method allows selective retrieval of different variables
+        (trainable, non-trainable, optimizer, and metrics) associated with the
+        model. The variables are returned in a nested dictionary format,
+        where the keys correspond to the variable names and the values are the
+        nested representations of the variables.
+
+        Args:
+            trainable_variables: Whether to include trainable variables.
+            non_trainable_variables: Whether to include non-trainable variables.
+            optimizer_variables: Whether to include optimizer variables.
+            metrics_variables: Whether to include metrics variables.
+
+        Returns:
+            dict: A dictionary containing the nested representations of the
+                requested variables. The keys are the variable names, and the
+                values are the corresponding nested dictionaries. If no variable
+                types are requested, an empty dictionary is returned.
+        """
+        variables = {}
+        if trainable_variables:
+            variables["trainable_variables"] = self._create_nested_dict(
+                self.trainable_variables
+            )
+        if non_trainable_variables:
+            variables["non_trainable_variables"] = self._create_nested_dict(
+                self.non_trainable_variables
+            )
+        if optimizer_variables:
+            variables["optimizer_variables"] = self._create_nested_dict(
+                self.optimizer.variables
+            )
+        if metrics_variables:
+            variables["metrics_variables"] = self._create_nested_dict(
+                self.metrics_variables
+            )
+
+        return variables
+
+    def _create_nested_dict(self, variables):
+        flat_dict = {}
+        for v in variables:
+            flat_dict[v.path] = v.value
```

Comment:
As we discussed, paths aren't unique. So this won't work. For instance if two layers have the same manually-set name in a subclassed model.

Either we use an order-based data structure where names are treated as metadata (e.g. list of tuples `(path, value)`) or we uniquify the paths in a silly-looking way by appending `(1)`, `(2)` etc to non-unique paths that encountered again

What is the reason for preferring a dict in this context?

Reply:
I'm using dict to keep the tree-like structure of the variables because that's helpful for debugging purposes. 
I can make the dictionary a mapping between paths and "list" of values instead so if there are non-unique paths, we can still restore them based on the order they are placed in the list, e.g. `dict[path]=[v1, v2, v3]`. How does that sound?

Reply:
As discussed offline, I changed this to raise an error if the path is duplicated

Reply:
I know this is resolved, but I do think a nested dict will fit much better with the expectations of anyone using Orbax. And might allow most custom workflows, like a partial restore from a checkpoint.

The main alternative to using variable.path that I can think of would be to use the same state walking mechanisms as our saving code. So the structure of the state dict is one-to-one with the structure of our `.weights.h5` files. But that seems significantly more complext.